### ISSUE TITLE
exiting the game using a keyboard shortcut will exit gamescope and not the game

### ISSUE THREAD

**Issue by beh-10257** [ValveSoftware/gamescope#885]
in my system to exit any application I just use SUPER+C the default for Hyprland
anyway so basically after pressing that key combination while running apex legends and after checking in htop
we can see that gamescope actually closes but all his child processes don't
aka the game
so yeah I guess the expected behaviour is to gamescope to kill his child processes

**Comment by MithicSpirit** [comment]
Can reproduce on sway as well. IMO gamescope should wait for all windows inside it to close before closing itself.

**Comment by Cloudef** [comment]
bumping

**Comment by Joshua-Ashton** [comment]
Gamescope does try to kill its children when it dies but if you are forcing that it's likely not getting a chance.

**Comment by MithicSpirit** [comment]
I don't think that it's forcing. I'm pretty sure that it's the equivalent of pressing the "x" button to close a window. For example, doing the same on a LibreOffice Writer window still allows it to prompt the user on whether to save the document or not.

**Comment by Cloudef** [comment]
I think this might actually be issue with children that double forks

**Comment by Cloudef** [comment]
Hmm, I tried make gamescope call my `reaper` binary from here https://github.com/Cloudef/local-daemon 
And the reaper indeed does not seem to receive any sigterm from gamescope.

Wonder if wlroots/hyprland is being too agressive in closing the process, or gamescope does not handle wayland disconnect properly? Here are some logs from the end:
```
(EE) failed to read Wayland events: Broken pipe
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
X connection to :1 broken (explicit kill or server shutdown).
```

**Comment by MithicSpirit** [comment]
> w/ SIGTERM being the signal for requesting an application to close

Are you sure? I've just tried to send SIGTERM to a LibreOffice Writer window and it just closed immediately without prompting to save the document (unlike when I close it via Sway).

**Comment by sharkautarch** [comment]
> > w/ SIGTERM being the signal for requesting an application to close
> 
> Are you sure? I've just tried to send SIGTERM to a LibreOffice Writer window and it just closed immediately without prompting to save the document (unlike when I close it via Sway).


programs are able to ignore SIGTERM signals, but not SIGKILL signals
https://www.gnu.org/software/libc/manual/html_node/Termination-Signals.html

**Comment by MithicSpirit** [comment]
> > > w/ SIGTERM being the signal for requesting an application to close
> > 
> > 
> > Are you sure? I've just tried to send SIGTERM to a LibreOffice Writer window and it just closed immediately without prompting to save the document (unlike when I close it via Sway).
> 
> programs are able to ignore SIGTERM signals, but not SIGKILL signals [gnu.org/software/libc/manual/html_node/Termination-Signals.html](https://www.gnu.org/software/libc/manual/html_node/Termination-Signals.html)

I just meant that SIGTERM is not what's used for an application's *window* to close. I feel like the solution to this issue would be for gamescope to forward the window close event to all windows inside it and stay open until those windows close, rather than just immediately closing the gamescope window and trying to stop the processes in some other way (like SIGTERM).

**Comment by MithicSpirit** [comment]
> Gamescope does try to kill its children when it dies but if you are forcing that it's likely not getting a chance.

Just got bit by this issue again, and decided to do some more testing. If I try to close native `xev` in sway, the last couple messages are
```
KeyPress event, serial 34, synthetic NO, window 0x400001,
    root 0x414, subw 0x0, time 103095070, (632,910), root:(640,963),
    state 0x40, keycode 50 (keysym 0xffe1, Shift_L), same_screen YES,
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

ClientMessage event, serial 34, synthetic YES, window 0x400001,
    message_type 0xf6 (WM_PROTOCOLS), format 32, message 0xf5 (WM_DELETE_WINDOW)
```
while doing so in gamescope instead ends with
```
KeyPress event, serial 32, synthetic NO, window 0x400001,
    root 0x416, subw 0x0, time 103121162, (168,168), root:(170,170),
    state 0x40, keycode 50 (keysym 0xffe1, Shift_L), same_screen YES,
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False
(EE) failed to read Wayland events: Broken pipe
[gamescopereaper] [Info]  reaper: Parent of gamescopereaper was killed. Killing children.
[gamescope] [Info]  launch: Primary child shut down!
```
notice the lack of a `WM_DELETE_WINDOW` message (the Shift_L is part of the keybind to close the window).

While the logs seem to suggest that gamescope connection was abruptly closed ("Broken pipe"), I don't believe this to be the case, as running native `wev` has it end on a `[10:    xdg_toplevel] close`, which is [documented](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/blob/main/stable/xdg-shell/xdg-shell.xml?ref_type=heads#L1137-L1148) as being an ignorable request. The fact that this is optional is further demonstrated by the fact that LibreOffice responds to such a command by opening a save dialog, rather than immediately closing.

The behavior I expect is that gamescope would respond to a `xdg_toplevel::close` message not by closing, but by sending a `WM_DELETE_WINDOW` message to the client that is currently in focus (or another `xdg_toplevel::close` for Wayland native clients). If clients not responding to this message is a worry, maybe gamescope should revert to the current behavior if it receives multiple requests to close in a short interval (though I would rather not—SIGKILL is always there in a pinch).

**Comment by MithicSpirit** [comment]
Further investigation leads me to believe that the culprit is the `LibDecor_Frame_Close` function, which makes gamescope close itself: https://github.com/ValveSoftware/gamescope/blob/7dd1bcd9102a17e039970ccd9a324a9fe8365d6d/src/Backends/WaylandBackend.cpp#L1161-L1164 I fear I may be in over my head at this point.

**Comment by sharkautarch** [comment]
> Further investigation leads me to believe that the culprit is the `LibDecor_Frame_Close` function, which makes gamescope close itself:
> 
> https://github.com/ValveSoftware/gamescope/blob/7dd1bcd9102a17e039970ccd9a324a9fe8365d6d/src/Backends/WaylandBackend.cpp#L1161-L1164
> I fear I may be in over my head at this point.


hmm I have no clue whether that function is guaranteed to only run if the **root** window is closed, or if that could also be run when a child window is closed.
Because, iirc, I believe that gamescope's current intended behavior when running in nested mode (aka "desktop mode") is to close itself once the root window is closed (tho first it attempts to ensure all child processes are cleaned up before exiting, via its custom signal handler for `SIGTERM` et al, and also with the parent gamescopereaper process) 

@misyltoad do you know if `CWaylandPlane::LibDecor_Frame_Close()` only gets called on the root window?

**Comment by MithicSpirit** [comment]
> hmm I have no clue whether that function is guaranteed to only run if the **root** window is closed, or if that could also be run when a child window is closed.

Being in the WaylandBackend and used with LibDecor, I believe that it is run if the gamescope window itself receives a close request from the host compositor, independent of what's happening *within* gamescope as a compositor.

> Because, iirc, I believe that gamescope's current intended behavior when running in nested mode (aka "desktop mode") is to close itself once the root window is closed (tho first it attempts to ensure all child processes are cleaned up before exiting, via its custom signal handler for `SIGTERM` et al, and also with the parent gamescopereaper process)

The cleanup code is a bit confusing (probably because of my lack of knowledge of X11 and Wayland), but I think that the relevant code is at the end of the `wlserver_run` function. I think part of the issue might be with how Wine/the game handles its window/Xwayland being destroyed.
https://github.com/ValveSoftware/gamescope/blob/7dd1bcd9102a17e039970ccd9a324a9fe8365d6d/src/wlserver.cpp#L1983-L1993

**Comment by MithicSpirit** [comment]
I'm trying to work on this, but I can't figure out how to send `WM_DELETE_WINDOW` to Xwayland clients. I tried using `wlr_xwayland_surface_close`, but that requires a `wlr_xwayland_surface`, which I've not been able to get (`wlr_xwayland_surface_try_from_wlr_surface` doesn't work, even using surfaces that come from a window with type `XWAYLAND`).

Any ideas or recommendations would be appreciated.
